pyfr/integrators/base.py

```diff
@@ -183,12 +167,36 @@ def _get_gndofs(self):
 
     @memoize
     def _get_axnpby_kerns(self, *rs, subdims=None):
+        kerns = [self.backend.kernel('axnpby', *[em[r] for r in rs],
+                                     subdims=subdims)
+                 for em in self.system.ele_banks]
+
+        return kerns
+
+    @memoize
+    def _get_errest_kerns(self, *rs, norm):
+        kerns = [self.backend.kernel('reduction', *[em[r] for r in rs],
+                                     method='errest', norm=norm)
+                 for em in self.system.ele_banks]
+
+        return kerns
+
+    @memoize
+    def _get_resid_kerns(self, rcurr, rold, norm):
         kerns = []
-        for em in self.system.ele_banks:
-            regs = [em[r] for r in rs]
-            kerns.append(self.backend.kernel(
-                'axnpby', *regs, subdims=subdims
-            ))
+        if hasattr(self, 'dtau_upts'):
```

Comment:
We should look to restructure things to avoid code duplication for these blocks.

Reply:
This bit bothers me as well. Do you have any suggestions? I thought about using enumerate and accessing dtaumats from dtau_upts array with an index but that wouldn't look good.

Reply:
My feeling is we want to get to the point where we always pass `dt_mat` but have the value clearly take on a value of None when we do not have such a matrix.